src/main/java/ca/warp7/frc2023/subsystems/Elevator.java

```diff
@@ -0,0 +1,121 @@
+    }
+
+    // makes a leader and a follower motor for the left side
+    public CANSparkMax LeftMain = createMasterSparkMAX(kElevatorLeftMain);
+    public CANSparkMax LeftSecondary = createFollowerSparkMAX(kElevatorLeftSecondary);
+
+    // makes the inverted follower motors for the right side
+    public CANSparkMax RightMain = createFollowerSparkMAXInverted(kElevatorRightMain);
+    public CANSparkMax RightSecondary = createFollowerSparkMAXInverted(kElevatorRightSecondary);
+
+    // Check if limit switch is forward or Reverse irl
+    // Check if we are going to use Normally closed or Normally open
+    // used to sync with encoders
+    public SparkMaxLimitSwitch limitSwitchLeftMain = LeftMain.getForwardLimitSwitch(Type.kNormallyClosed);
+
+    // Creates an incremental encoder
+    public RelativeEncoder encoder = LeftMain.getAlternateEncoder(SparkMaxAlternateEncoder.Type.kQuadrature, 8192);
+    // Method that creates and sets up a PID controller
+    public SparkMaxPIDController createAndSetupPIDController(CANSparkMax motor) {
+        SparkMaxPIDController PIDController = motor.getPIDController();
+        PIDController.setFeedbackDevice(encoder);
+        double kP = 0.1;
+        double kI = 1e-4;
+        double kD = 1;
+        double kIz = 0;
+        double kFF = 0;
+        double kMaxOutput = 1;
+        double kMinOutput = -1;
+        int smartMotionSlot = 0;
+        double kMaxVel = 2000;
+        double kMinVel = 0;
+        double kMaxAcc = 1500;
+        PIDController.setP(kP);
+        PIDController.setI(kI);
+        PIDController.setD(kD);
+        PIDController.setIZone(kIz);
+        PIDController.setFF(kFF);
+        PIDController.setOutputRange(kMinOutput, kMaxOutput);
+        PIDController.setReference(0, CANSparkMax.ControlType.kSmartMotion);
+        PIDController.setSmartMotionMaxVelocity(kMaxVel, smartMotionSlot);
+        PIDController.setSmartMotionMinOutputVelocity(kMinVel, smartMotionSlot);
+        PIDController.setSmartMotionMaxAccel(kMaxAcc, smartMotionSlot);
+        PIDController.setSmartMotionAllowedClosedLoopError(kMinOutput, smartMotionSlot);
+        return PIDController;
+    }
+
+    // creates and sets a PID controller
+    public SparkMaxPIDController controllerLeftMain = createAndSetupPIDController(LeftMain);
+
+    // gets the number of rotations from the encoder
+    public double getPosition() {
+        return encoder.getPosition();
+    }
+
+    // Sets the setPoint of the PID controller to whatever was passed as the input
+    public void setPosition(double length) {
+        controllerLeftMain.setReference(length, CANSparkMax.ControlType.kPosition);
+    }
+
+    // resets the encoder
+    public void zeroEncoder() {
+        if (limitSwitchLeftMain.isPressed()) {
+            encoder.setPosition(0);
+        } else {
+            System.out.println("limit switch is no pressy");
+        }
+    }
+
+    // fully retracts elevator
+    public void fullyRetract() {
```

Comment:
I'm confused by the flow of logic. Calling fullyRetract() calls zeroEncoder(). Within the method, it checks if the limit switch is pressed, which it shouldn't be unless you are already fully retracted, thus never actually zeroing the encoder.

Reply:
@Colepng

Reply:
The idea here is it fully retracts and then zero the encoder but I can't figure out a way to wait untill limit switch is pressed to zero the encoder, normally you would do a while loop but can't do that with frc

Reply:
It's one of the things that makes robot programming hard  (and game programming, too, which was my first exposure to this paradigm)

The first robotics infrastructure usually calls an update about 50 times a second.  Typically, robotics code does stuff that it would normally do in the while loop during the update code.  The update call should be non-blocking, which is a fancy way of saying "do your stuff and exit as fast as possible"

My understanding (remembering that I've never been lucky enough to program a competition robot) is that sub-systems can defined an update function that'll be called 50 times a second.  The update function could look something like this...

update() {
  if ( mode == RETRACTING ) {
     if ( limitSwitchActivated() ) {
        resetEncoder();       // we're at zero if the limit switch is activated
        mode = NORMAL;  // go back to normal operation
     }
     else {
        retractMore(); // definition TBD, but it has to move the elevator more toward the limit
     }
  }
}

One thought I'll pitch - if the elevator starts at zero before the competition, you might not need this level of code complexity during the match.  Like maybe in the pits there's a code option to retract the elevator, but once the competition starts the elevator is assumed to be retracted and zeroed.  It's one less thing to worry about for competition code.